### PR TITLE
Nonuniform BoundaryConditions

### PR DESCRIPTION
**Pull request by dschwoerer**
Add boundaries which are aware of non-uniform grid spacing:

* Dirichlet
* Neumann
* Free
* for 2nd, 3rd and 4th each

~Tests are not included, as they require `realy`~

The code uses sympy to generate the C++ code, so extend should be easy.

### CODE REVIEW

**Comment by johnomotani** [comment]
Definitely agree there is an issue here.

I think from a skim of your code here that the solution you implement is taking into account explicitly the positions of each point in x-y space. That, if I follow these lecture notes correctly --- https://www.nada.kth.se/kurser/kth/2D1263/l6.pdf [see section 4.3, in particular eqs (4.2), (4.4) and figs 3.3 and 3.5-6] --- is the most accurate thing to do, but our derivative operators don't do that, they just use the local grid spacing at the location of the derivative (updating the non-uniform grid support in the derivatives could be a useful thing to do...?).

Using the local spacing at the boundary is a bit of a pain, especially with staggered fields (they end up needing the grid spacing at the x-y corner of cells to put the x-boundary condition on a y-staggered field or vice versa). I have put together some changes, building on other stuff I was doing to get difops to converge accurately near boundaries, that get the grid spacings at the boundary locations. See the derivatives3 test in e8b28bbfed449077f34b723fbd47707b2809f8f3, which is updated to have a non-constant grid spacing and to apply boundary conditions. With the updates, it converges correctly for dirichlet, neumann and free boundary conditions, on staggered and unstaggered grids. The cost compared to your approach is having to change a lot more stuff in Coordinates, essentially to get a `calcGridSpacing` in `src/mesh/boundary_op.cxx` to be possible. (Although many of the changes were to enable the 'x-boundary of y-staggered field' case which should work but I haven't actually managed to test yet.) Some of the changes I think could be good anyway, from a neatness and/or performance viewpoint, but that's a question for a different PR (the commit history and tests definitely need tidying up).

**Comment by dschwoerer** [comment]
Some timings: - based on 41382ba6f21b13d725bf032a9de9af339aaea399
```
dirichlet_o2:   10000 iterations took 0.776276 s
neumann_o2:     10000 iterations took 2.3407 s
free_o2:        10000 iterations took 1.56801 s
dirichlet_nu_o2:        10000 iterations took 0.613323 s
neumann_nu_o2:  10000 iterations took 0.614806 s
free_nu_o2:     10000 iterations took 0.649469 s
dirichlet_o3:   10000 iterations took 5.79157 s
free_o3:        10000 iterations took 1.76174 s
dirichlet_nu_o3:        10000 iterations took 0.792008 s
neumann_nu_o3:  10000 iterations took 0.78357 s
free_nu_o3:     10000 iterations took 0.846108 s
dirichlet_o4:   10000 iterations took 6.96777 s
dirichlet_nu_o4:        10000 iterations took 1.02474 s
neumann_nu_o4:  10000 iterations took 1.54527 s
free_nu_o4:     10000 iterations took 1.06533 s
```

Only `dirichlet_o4` is slightly faster then the more general versions. Thus ether the general versions could be improved (Is #1351 making them faster then the non-uniform versions?) or the more general version could be used as a default. If there is no performance penalty - I think we should try to do "the right thing"

I thought of implementing some adaptive stencils - but haven't came around to do it. I thought of doing it as an aiolos-only feature, as the standard stencil code does not know the position of the derivative.

**Comment by dschwoerer** [comment]
This is ready to be merged - can someone have a look over this?

As this is to my understanding the correct implementation, thus we could discuss to replace the old, non-uniform mesh aware implementation, but I thought as a first step we could have it in parallel, to allow testing ...

File: src/mesh/boundary_nonuniform.cxx [hunk 1]
```diff
@@ -0,0 +1,1285 @@
...
+
+  BoutReal val = 0.0;
+
+  Mesh* mesh = f.getMesh();
+  CELL_LOC loc = f.getLocation();
+
+  BoutReal vals[mesh->LocalNz];
+  int bx = bndry->bx;
+  int by = bndry->by;
+  int stag = 0;
```
Comment (ZedThree) [inline src/mesh/boundary_nonuniform.cxx]: Is this the signed-direction of the stagger? Does it ever make sense to take on other values? It might make sense to make this be an `enum class`

Maybe also `stag` -> `stagger`

File: include/field3d.hxx [hunk 2]
```diff
@@ -295,6 +295,13 @@ class Field3D : public Field, public FieldData {
     return data[d.ind];
   }
 
+  BoutReal& operator[](const Indices &d) {
```
Comment (ZedThree) [inline include/field3d.hxx]: I'd like to avoid bringing these operators back if possible. If we added something like

```cpp
// const Mesh& would be better, but I think Mesh const-ness might need fixing
Ind3D indexFromXYZ(int x, int y, int z, Mesh* mesh)
```

would that avoid the need for `Indices`? If not, then I would suggest just moving `Indices` entirely into `boundary_nonuniform.cxx` and just making it POD:

```cpp
struct Indices {
  int x, y, z;
};
```

and explicitly indexing the fields with `f(i.x, i.y, i.z)`

File: src/mesh/boundary_nonuniform.cxx [hunk 3]
```diff
@@ -0,0 +1,1285 @@
...
+  }
+  if (loc == CELL_YLOW) {
+    if (by == 0) {
+      by = -1;
+    } else if (by < 0) {
+      stag = -1;
+    } else {
+      stag = 1;
+    }
+  }
```
Comment (ZedThree) [inline src/mesh/boundary_nonuniform.cxx]: These two conditionals are repeated in each `apply`. Can they be pulled out into free function(s)?

File: src/mesh/boundary_nonuniform.cxx [hunk 4]
```diff
@@ -0,0 +1,1285 @@
...
+    if (stag == 0) {
+      x0 = 0;
+      BoutReal st = 0;
+      t = dy[i1];
+      x1 = st + t / 2;
+      st += t;
+    } else {
+      x0 = 0; // dy(bndry->x, bndry->y) / 2;
+      x1 = x0 + dy[i1];
+    }
```
Comment (ZedThree) [inline src/mesh/boundary_nonuniform.cxx]: This code is also repeated a lot -- can it be pulled out?

When we eventually get the boundary stuff into a framework like the derivatives, this bit will be something like `populateStencil`. Can we do something like that now? Put `x0`, `x1` and `fac0`, `fac1` into separate stencils/structs?

File: src/mesh/boundary_nonuniform.cxx [hunk 5]
```diff
@@ -0,0 +1,1285 @@
...
+  if (!args.empty()) {
+    // First argument should be an expression
+    newgen = FieldFactory::get()->parse(args.front());
+  }
+  return new BoundaryDirichletNonUniform_O2(region, newgen);
+}
+
+void BoundaryDirichletNonUniform_O2::calc_interp_to_stencil(BoutReal x0, BoutReal x1,
+                                                            BoutReal& fac0,
+                                                            BoutReal& fac1) const {
```
Comment (ZedThree) [inline src/mesh/boundary_nonuniform.cxx]: If `fac0`, `fac1` were in a struct, they could be a return value rather than out parameters

**Review by ZedThree — CHANGES_REQUESTED** [review]
Looking good. I'd like to get this in so we can base the boundary rewrite on it.

**Comment by ZedThree** [comment]
It would be really good if this included the tests as well. Can they be written without using `realy`?

File: include/field3d.hxx (hunk 2 above)
Comment (dschwoerer) [reply include/field3d.hxx]: I like Indices because they allow better readability, and are also not worse then integer indices performance wise. Further they allow to write code for Fields. Finally reintroducing them removes a breaking change from the 4.2(?) minor release.

I did not understand why they were removed in the first place, could you explain?

Comment (ZedThree) [reply include/field3d.hxx]: We removed them because we don't want to have to maintain many different ways of indexing fields. Also `Indices` was really an implementation detail of `DataIterator`. Now that we have improved on `DataIterator` it doesn't make sense to keep `Indices` about as well.

**Comment by johnomotani** [comment]
@dschwoerer could you add a docstring explaining how the coefficients are calculated? I'd find that very helpful.

**Comment by dschwoerer** [comment]
Do you mean the derivation, starting with the taylor expansion, or rather explain the implementation?

**Comment by johnomotani** [comment]
If you could sketch the derivation, with notation corresponding to the variables in the code, that would be great :1st_place_medal:

File: src/mesh/boundary_nonuniform.cxx.in.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    if (by == 0) {
+      by = -1;
+    } else if (by < 0) {
+      stag = -1;
+    } else {
+      stag = 1;
+    }
+  }
+{% if what == "Dirichlet" %}
+  int istart = (stag == -1) ? -1 : 0;
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx.in.py]: I don't quiet remember why this is not the case.
But I have a test, and according to that, this is correct :-)

**Comment by dschwoerer** [comment]
I think I should have addressed all comments. Sorry for the delay, but I stopped using non-uniform meshes for a while ...

File: src/mesh/boundary_nonuniform.cxx (hunk 4 above)
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx]: There are slight differences, depending on what BC is used. Thus I kept it for now.
The more trivial for loop has been removed.

**Comment by dschwoerer** [comment]
Can someone give the failed build a kick?

@johnomotani is the description in bfc4cbc sufficient?

**Review by ZedThree — CHANGES_REQUESTED** [review]
LGTM, just needs some polish in places. Basically, move local variables to the smallest possible scope and make them `const` where possible.

Also, by templating a few things now, this should make converting everything to templates later easier.

File: src/mesh/stencils_sympy.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+def gen_code(order,matrix):
+    global x
+    x=[]
+    for i in range(order):
+        x.append(Symbol("spacing.f%d"%i))
+    A=Matrix(order,order,matrix)
+
+    try:
+        iA=A.inv()
+    except:
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/stencils_sympy.py]: Why not?

from [here](https://www.tutorialspoint.com/python3/python_exceptions.htm)
This kind of a try-except statement catches all the exceptions that occur. Using this kind of try-except statement is not considered a good programming practice though, because it catches all exceptions but does not make the programmer identify the root cause of the problem that may occur.

However, the exception is raised as is. This is not worse then not try/except.
It however gives some infos, that may or may not be useful.

File: src/mesh/stencils_sympy.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+def fac(a):
+    if a==0 or a==1:
+        return 1
+    else:
+        assert(a>0)
+        return a*fac(a-1)
+
+def gen_code(order,matrix):
+    global x
+    x=[]
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/stencils_sympy.py]: I am confused.
Isn't then taylor and x global?

Also, Taylor has a copy of x, but still uses the global one (though that is easy to fix) ...

Assuming that shouldn't be at global scope.
taylor is an argument to the function, so cannot be created within the function, unless I build also a factory ...

The easiest way would be to make neuman etc objects, rather then functions. Then we can `setX(x)` on them after they are passed in ...

I can change it, sure, but not sure it is much nicer ...

[the remark this answers is not in the document]
Comment (ZedThree) [reply src/mesh/stencils_sympy.py]: Yes, sorry, that was just showing the basic idea.

You can pass a class name as the second argument to `gen_code` as something to create:

```python
class Taylor:
   # as above
class Neumann:
   # similar
```

def gen_code(order, matrix_type):
    x = [sympy.Symbol("spacing.f{:1d}".format(i)) for i in range(order)]
    matrix = matrix_type(x)
    A = sympy.Matrix(order, order, matrix)
    ...

gen_code(3, Taylor)
```

`matrix_type` just needs to return a callable that takes two arguments. It's a pretty small change:

```diff
- def neumann(i,j):
+ class Neumann:
+     def __init__(self, x):
+          self.taylor = Taylor(x)
+     def __call__(self, i, j):
```

(near enough)

File: src/mesh/boundary_nonuniform.cxx.in.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx.in.py]: I did remember. Added a comment.

File: src/mesh/boundary_nonuniform.hxx.in.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+struct vec{{order}}{
+{% for i in range(order) %}
+  BoutReal f{{i}};
+{% endfor %}
+  void operator+=(BoutReal v){
+{% for i in range(order) %}
+    f{{i}} += v;
+{% endfor %}
+  }
+};
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.hxx.in.py]: `+=` is much nicer, imho.

Also, is there a major advantage of using an array? It might interfere with optimisations ...

File: tests/MMS/non-uniform-boundaries/runtest [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+    ["free_nu_o2", np.cos,2],
+    ["free_nu_o3", np.cos,[4,2]],
+    ["free_nu_o4", np.cos,[4,2]],
+    ["free_nu_o2", np.sin,[3,1]],
+    ["free_nu_o3", np.sin,[3,1]],
+    ["free_nu_o4", np.sin,[5,1]]
+]
+
+code=0
+locs=["CENTRE", "YLOW" ]#"XLOW", "YLOW", "ZLOW":]
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply tests/MMS/non-uniform-boundaries/runtest]: added.

It might make sense to extend this a bit, to take variations in x also into account.

Also running a test in x direction would be useful ...

File: src/mesh/stencils_sympy.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/stencils_sympy.py]: Yes, make sense :+1: 

I was confused about python allowing to pass classes, rather then objects around.

File: src/mesh/stencils_sympy.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+def pow(a,b):
+    if b == 0:
+        return "1"
+    if b == 1:
+        return a
+    else:
+        return "%s**%d"%(a,b)
+
+def fac(a):
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/stencils_sympy.py]: sure, but didn't know about it. Was easier to implement then to look it up.

the function is in [sympy.functions.combinatorial.factorials.factorial](
https://docs.sympy.org/latest/modules/functions/combinatorial.html?highlight=factorial#sympy.functions.combinatorial.factorials.factorial)
Implementing it is easier then to remember this name ...

File: src/mesh/stencils.md [hunk 12]
```diff
@@ -0,0 +1,29 @@
+Notes concerning the generation of stencils
+================
+
+We want to create a Tayler function
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/stencils.md]: thanks, fixed

File: src/mesh/boundary_nonuniform.hxx.in.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+header="""
+#include <utility>
+
+#include "boundary_op.hxx"
+
+// Define structs used
+struct Indices{
+  int x,y,z;
+};
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.hxx.in.py]: I used to pass it around, but then you complained. So that is why it is here.
I think having them in a struct is clarer then having just several integers.

File: src/mesh/boundary_factory.cxx [hunk 14]
```diff
@@ -12,6 +12,8 @@ using std::string;
 
 #include <output.hxx>
 
+#include "boundary_nonuniform.hxx"
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_factory.cxx]: done

File: src/mesh/boundary_nonuniform.cxx.in.py [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+        }
+        facs = calc_interp_to_stencil(spacing);
+        if (stagger == 1) {
+          spacing += t;
+        }
+      }
+      for (ic.z = 0; ic.z < mesh->LocalNz; ic.z++) {
+{% for i in range(1,order) %}
+        i{{i}}.z = ic.z;
+{% endfor %}
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx.in.py]: Fixed

File: src/mesh/boundary_nonuniform.cxx.in.py [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+                                + mesh->GlobalY(bndry->y - y_boundary_offset)); // the grid cell
+
+        vals[zk] = fg->generate(bout::generator::Context().set("x", xnorm, "y", TWOPI * ynorm, "z", TWOPI * zk / (mesh->LocalNz), "t" , t));
+      }
+    }
+{% endif %}
+
+
+    vec{{order}} spacing;
+    vec{{order}} facs;
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx.in.py]: yes, done

File: src/mesh/boundary_nonuniform.cxx.in.py [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+  BoutReal val = 0.0;
+  BoutReal vals[mesh->LocalNz];
+{% endif %}
+
+  int x_boundary_offset = bndry->bx;
+  int y_boundary_offset = bndry->by;
+  int stagger = 0;
+  update_stagger_offsets(x_boundary_offset, y_boundary_offset, stagger, loc);
+{% if type == "Dirichlet" %}
+  int istart = (stagger == -1) ? -1 : 0;
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx.in.py]: thanks, done

File: src/mesh/boundary_nonuniform.cxx.in.py [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+    vec{{order}} spacing;
+    vec{{order}} facs;
+
+    const Field2D &coords_field =
+        bndry->by != 0 ? mesh->getCoordinates()->dy : mesh->getCoordinates()->dx;
+{% if type != "Free" %}
+{% for i in range(1,order) %}
+    Indices i{{i}}{bndry->x - {{i}} * bndry->bx, bndry->y - {{i}} * bndry->by, 0};
+{% endfor %}
+    BoutReal t;
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx.in.py]: yes, t is for temporary :-)
Renamed + reduced scope

File: src/mesh/boundary_nonuniform.cxx.in.py [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+  // Decide which generator to use
+  std::shared_ptr<FieldGenerator> fg = gen;
+  if (!fg)
+    fg = f.getBndryGenerator(bndry->location);
+
+  Mesh *mesh = f.getMesh();
+  CELL_LOC loc = f.getLocation();
+
+{% if type != "Free" %}
+  BoutReal val = 0.0;
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx.in.py]: done

File: src/mesh/boundary_nonuniform.cxx.in.py [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+  BoutReal vals[mesh->LocalNz];
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx.in.py]: done

File: src/mesh/boundary_nonuniform.cxx.in.py [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+#include <fft.hxx>
+#include <globals.hxx>
+#include <invert_laplace.hxx>
+#include <msg_stack.hxx>
+#include <output.hxx>
+#include <utils.hxx>
+
+#include "boundary_nonuniform.hxx"
+
+static void update_stagger_offsets(int& x_boundary_offset, int& y_boundary_offset, int& stagger, CELL_LOC loc){
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx.in.py]: I could get used to the C++17 version, assuming I can do something like:
```
const auto [y_boundary_offset, stagger] = offset<CELL_YLOW>(bndry->by, stagger, loc);
```
I.e. update stagger.

The other version requires to much handling of extracting stagger, to so that the function is ~6 lines of code, and the calling side requires ~ 4.

In that case I would rather put it directly in the code, then call a function.
For the generated code, this isn't really an issue ...

File: examples/performance/boundary/boundary.cxx [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+  std::string name;
+  opt->getSection("f")->get("bndry_all",name,"error");
+
+  double time = Timer::getTime("dirichlet_o3");
+  printf("%-20s | %12.8f s", name.c_str(), time);
+  
+  return 0;
+}
+
+BOUTMAIN(BoundaryOp_timing);
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply examples/performance/boundary/boundary.cxx]: done

File: src/mesh/boundary_nonuniform.cxx [hunk 23]
```diff
@@ -0,0 +1,1389 @@
...
+      stagger = -1;
+    } else {
+      stagger = 1;
+    }
+  }
+}
+
+#if !BOUT_USE_METRIC_3D
+using IndMetric = Ind2D;
+#define IND(var, x, y, z) IndMetric var(x* localNy + y, localNy, 1)
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx:45]: I could use a jinja2 macros instead of a C macros. Would that be preferred? Certainly clang-tidy would be happy ...

**Comment by dschwoerer** [comment]
| Boundary condition | 1000 ; 2D ; 128^3 |	 1000 ; 2D ; 128x16384x1 |	  1000 ; 3D ; 128^3 |
| ----------- | -----------: |	 -----------: |	 -----------: |
| dirichlet_o2 | 0.353929 s |	 0.476291 s |	 0.375638 s |
| neumann_o2 | 0.442191 s |	 0.5878 s |	 0.856184 s |
| free_o2 | 0.217463 s |	 0.421346 s |	 0.243244 s |
| dirichlet_nu_o2 | 0.187466 s |	 1.97744 s |	 2.78296 s |
| neumann_nu_o2 | 0.177839 s |	 1.79646 s |	 2.10596 s |
| free_nu_o2 | 0.187784 s |	 2.74089 s |	 3.44364 s |
| dirichlet_o3 | 0.829212 s |	 0.785762 s |	 0.907647 s |
| free_o3 | 0.309709 s |	 0.984481 s |	 0.382664 s |
| dirichlet_nu_o3 | 0.251681 s |	 3.85593 s |	 5.58085 s |
| neumann_nu_o3 | 0.238788 s |	 2.94091 s |	 3.49203 s |
| free_nu_o3 | 0.287577 s |	 3.22091 s |	 4.60549 s |
| dirichlet_o4 | 1.09995 s |	 0.773845 s |	 1.07976 s |
| dirichlet_nu_o4 | 0.339356 s |	 3.487 s |	 4.87303 s |
| neumann_nu_o4 | 0.362675 s |	 4.04734 s |	 6.26752 s |
| free_nu_o4 | 0.386731 s |	 3.74004 s |	 5.49017 s |


So for 2D and large enough `nz` the overhead of the correct stencil is minimal. Otherwise it can be expensive. Would be nice to know why 3D is slower, probably something that can be fixed.

File: src/mesh/boundary_nonuniform.cxx [hunk 24]
```diff
@@ -0,0 +1,3324 @@
...
+                * (mesh->GlobalY(bndry->y)                         // In the guard cell
+                   + mesh->GlobalY(bndry->y - y_boundary_offset)); // the grid cell
+            const BoutReal zfac = TWOPI / mesh->LocalNz;
+            for (int zk = 0; zk < mesh->LocalNz; zk++) {
+              vals[zk] = fg->generate(bout::generator::Context().set(
+                  "x", xnorm, "y", ynorm, "z", zfac * zk, "t", t));
+            }
+          }
+
+          vec2 spacing;
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx:402]: I would like to avoid the overhead of that ... (applies to all copies of the uninitialised warnings)

File: src/mesh/boundary_nonuniform.cxx [hunk 25]
```diff
@@ -0,0 +1,3324 @@
...
+                                            }
+                                            facs = calc_interp_to_stencil(spacing);
+
+#if !BOUT_USE_METRIC_3D
+                                            for (int iz{0}; iz < mesh->LocalNz; iz++) {
+#endif
+#if BOUT_USE_METRIC_3D
+#define MAKE3D(x) x
+#else
+#define MAKE3D(x) x##3d
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx:2031]: I doubt that can be replaced with a template function ...
Is it possible?

[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/boundary_nonuniform.cxx:2031]: I guess something like this should work:
```
struct Make3D {
  const FieldMetric& fm;
#if BOUT_USE_METRIC_3D
  const Field3D& f3d;
#else
  const Field3D f3d;
#endif
  Make3D(const FieldMetric& f) : fm(f), f3d(f) {};
}

**Comment by dschwoerer** [comment]
This is, as written, currently not compatible with 3D metrics. Fixing it requires some time, which I currently do not have.
